drivers/all_neighbors_system_dashboard/app/models/all_neighbors_system_dashboard/demographic_ratio_calculations.rb

```diff
@@ -0,0 +1,128 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module AllNeighborsSystemDashboard
+  module DemographicRatioCalculations
+    extend ActiveSupport::Concern
+
+    included do
+      def categories
+        @categories ||= HudUtility.races(multi_racial: true).values.map do |race|
+          HudUtility.ethnicities.values.map do |ethnicity|
+            [race, ethnicity]
+          end
+        end.flatten(1)
+      end
+
+      def compute_project_ratios
+        project_ids_from_groups = GrdaWarehouse::Hud::Project.where(id: filter.effective_project_ids_from_secondary_project_groups).pluck(:project_id)
+        results = project_ids_from_groups.map do |project_id|
+          by_project = universe.members.where(a_t[:project_id].eq(project_id))
+          next unless by_project.exists?
+
+          project_type = by_project.first.universe_membership.project_type
+          project_name = by_project.first.universe_membership.project_name
+
+          clients_by_segment = by_project.pluck(:client_id).uniq.count
+          households_by_segment = by_project.where(a_t[:relationship].eq('SL')).pluck(:client_id).uniq.count
+
+          categories.map do |race, ethnicity|
+            by_category = by_project.where(a_t[:primary_race].eq(race).and(a_t[:ethnicity].eq(ethnicity)))
+            next if by_category.count.zero?
+
+            clients_by_demographic = by_category.pluck(:client_id).uniq.count
+            households_by_demographic = by_category.where(a_t[:relationship].eq('SL')).pluck(:client_id).uniq.count
+
+            OpenStruct.new(
+              project_type: project_type,
+              project_id: project_id,
+              project_name: project_name,
+              primary_race: race,
+              ethnicity: ethnicity,
+              clients_by_demographic: clients_by_demographic,
+              clients_by_segment: clients_by_segment,
+              households_by_demographic: households_by_demographic,
+              households_by_segment: households_by_segment,
+              project_category: 'Project',
+              report_start: filter.start_date,
+              report_end: filter.end_date,
+            )
+          end.compact
+        end.compact
```

Comment:
Doesn't this mean we don't include categories that are empty?  I'd expect Tableau would be unhappy with that.  Maybe we should still include those?

Reply:
That is how the existing file is formatted, so I believe it is correct.